// flow-coverage-report/dist/lib/flow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectFlowCoverageForFile = exports.checkFlowStatus = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var checkFlowStatus = exports.checkFlowStatus = function () {
  var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(flowCommandPath, projectDir, tmpDirPath) {
    var tmpFilePath, res, statusData, unexpectedException;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            tmpFilePath = void 0;


            if (process.env.VERBOSE && process.env.VERBOSE === 'DUMP_JSON') {
              tmpFilePath = _temp2.default.path({ suffix: '.json', dir: tmpDirPath });
            }

            _context.next = 4;
            return (0, _promisified.exec)(flowCommandPath + ' status --json', { cwd: projectDir, maxBuffer: Infinity }, { dontReject: true });

          case 4:
            res = _context.sent;

            if (!(res.err && res.err.code !== 2)) {
              _context.next = 8;
              break;
            }

            if (process.env.VERBOSE) {
              console.error('Flow status error', res.err, res.stderr, res.stdout);
            }

            throw res.err;

          case 8:
            statusData = void 0;

            if (!tmpFilePath) {
              _context.next = 13;
              break;
            }

            _context.next = 12;
            return (0, _promisified.writeFile)(tmpFilePath, res.stdout || '');

          case 12:
            console.log('Flow status result saved to', tmpFilePath);

          case 13:
            _context.prev = 13;

            statusData = JSON.parse(String(res.stdout));
            _context.next = 22;
            break;

          case 17:
            _context.prev = 17;
            _context.t0 = _context['catch'](13);
            unexpectedException = _context.t0;

            // Verify the integrity of the format of the JSON status result.

            if (!unexpectedException) {
              _context.next = 22;
              break;
            }

            throw new Error('Parsing error on Flow status JSON result: ' + _context.t0);

          case 22:
            if (!(statusData && statusData.flowVersion)) {
              _context.next = 24;
              break;
            }

            return _context.abrupt('return', statusData);

          case 24:
            throw new Error('Invalid Flow status JSON format');

          case 25:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[13, 17]]);
  }));

  return function checkFlowStatus(_x, _x2, _x3) {
    return _ref2.apply(this, arguments);
  };
}();

// Definitions and flow types related to collectFlowCoverageForFile.

var collectFlowCoverageForFile = exports.collectFlowCoverageForFile = function () {
  var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(flowCommandPath, flowCommandTimeout, projectDir, filename, tmpDirPath) {
    var tmpFilePath, emptyCoverageData, res, parsedData, flowCoverageParsingError;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            tmpFilePath = void 0;


            if (process.env.VERBOSE && process.env.VERBOSE === 'DUMP_JSON') {
              tmpFilePath = _temp2.default.path({ suffix: '.json', dir: tmpDirPath });
            }

            emptyCoverageData = {
              filename: filename,
              annotation: 'no flow',
              expressions: {
                covered_count: 0,
                uncovered_count: 0,
                uncovered_locs: []
              }
            };


            if (process.env.VERBOSE) {
              console.log('Collecting coverage data from ' + filename + ' (timeouts in ' + flowCommandTimeout + ')...');
            }

            _context2.next = 6;
            return (0, _promisified.exec)(flowCommandPath + ' coverage --json ' + escapeFileName(filename),
            // NOTE: set a default timeouts and maxButter to Infinity to prevent,
            // misconfigured projects and source files that should raises errors
            // or hangs the flow daemon to prevent the coverage reporter to complete
            // the data collection. (See https://github.com/rpl/flow-coverage-report/pull/4
            // and https://github.com/rpl/flow-coverage-report/pull/5 for rationale,
            // thanks to to @mynameiswhm and @ryan953  for their help on hunting down this issue)
            { cwd: projectDir, timeout: flowCommandTimeout, maxBuffer: Infinity }, { dontReject: true });

          case 6:
            res = _context2.sent;

            if (!res.err) {
              _context2.next = 14;
              break;
            }

            console.error('ERROR Collecting coverage data from ' + filename + ' ', filename, res.err, res.stderr);

            if (!process.env.VERBOSE) {
              _context2.next = 13;
              break;
            }

            if (!tmpFilePath) {
              _context2.next = 13;
              break;
            }

            _context2.next = 13;
            return (0, _promisified.writeFile)(tmpFilePath, res.stdout || '');

          case 13:
            return _context2.abrupt('return', (0, _extends3.default)({}, emptyCoverageData, {
              percent: NaN,
              isError: true,
              flowCoverageError: undefined,
              flowCoverageException: res.err && res.err.message,
              flowCoverageStderr: res.stderr,
              flowCoverageParsingError: undefined
            }));

          case 14:
            if (!process.env.VERBOSE) {
              _context2.next = 20;
              break;
            }

            console.log('Collecting coverage data from ' + filename + ' completed.');

            if (!tmpFilePath) {
              _context2.next = 20;
              break;
            }

            _context2.next = 19;
            return (0, _promisified.writeFile)(tmpFilePath, res.stdout || '');

          case 19:
            console.log('Saved json dump of collected coverage data from ' + filename + ' to ' + tmpFilePath + '.');

          case 20:
            parsedData = void 0;
            flowCoverageParsingError = void 0;


            if (res.stdout) {
              try {
                parsedData = JSON.parse(String(res.stdout));
              } catch (err) {
                flowCoverageParsingError = err.message;
              }
            }

            if (res.stderr) {
              try {
                parsedData = JSON.parse(String(res.stderr));
                delete res.stderr;
              } catch (err) {}
            }

            if (!(parsedData && !parsedData.error)) {
              _context2.next = 30;
              break;
            }

            parsedData.filename = filename;
            _context2.next = 28;
            return (0, _flowAnnotationCheck.genCheckFlowStatus)(flowCommandPath, filename);

          case 28:
            parsedData.annotation = _context2.sent;
            return _context2.abrupt('return', parsedData);

          case 30:
            return _context2.abrupt('return', (0, _extends3.default)({}, emptyCoverageData, {
              percent: NaN,
              isError: true,
              flowCoverageError: parsedData && parsedData.error,
              flowCoverageException: undefined,
              flowCoverageParsingError: flowCoverageParsingError,
              flowCoverageStderr: res.stderr
            }));

          case 31:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function collectFlowCoverageForFile(_x4, _x5, _x6, _x7, _x8) {
    return _ref3.apply(this, arguments);
  };
}();

// Definition and flow types related to collectForCoverage.

exports.escapeFileName = escapeFileName;
exports.getCoveredPercent = getCoveredPercent;
exports.summarizeAnnotations = summarizeAnnotations;
exports.collectFlowCoverage = collectFlowCoverage;

var _minimatch = require('minimatch');

var _minimatch2 = _interopRequireDefault(_minimatch);

var _temp = require('temp');

var _temp2 = _interopRequireDefault(_temp);

var _flowAnnotationCheck = require('flow-annotation-check');

var _promisified = require('./promisified');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Load the Array.prototype.find polyfill if needed (e.g. nodejs 0.12).
/* istanbul ignore if  */
if (!Array.prototype.find) {
  require('array.prototype.find').shim();
}

// Escape special characters in file names.
function escapeFileName(fileName) {
  return fileName.replace(/(["\s'$`\\])/g, '\\$1');
}

/* eslint-disable camelcase */
function getCoveredPercent(_ref) {
  var covered_count = _ref.covered_count,
      uncovered_count = _ref.uncovered_count;

  var total = covered_count + uncovered_count;

  if (total === 0) {
    return 100;
  }

  return Math.floor(covered_count / total * 100);
}
/* eslint-disable-line camelcase */

// Definitions and flow types related to checkFlowStatus.

function summarizeAnnotations(coverageSummaryData) {
  var flowFiles = 0;
  var flowWeakFiles = 0;
  var noFlowFiles = 0;

  var filenames = Object.keys(coverageSummaryData.files);

  filenames.forEach(function (filename) {
    switch (coverageSummaryData.files[filename].annotation) {
      case 'flow':
        flowFiles += 1;
        break;
      case 'flow weak':
        flowWeakFiles += 1;
        break;
      case 'no flow':
        noFlowFiles += 1;
        break;
      default:
        throw new Error('Unexpected missing flow annotation on ' + filename);
    }
  });

  return {
    passed: flowWeakFiles + noFlowFiles === 0,
    flowFiles: flowFiles,
    flowWeakFiles: flowWeakFiles,
    noFlowFiles: noFlowFiles,
    totalFiles: filenames.length
  };
}

function collectFlowCoverage(flowCommandPath, flowCommandTimeout, projectDir, globIncludePatterns, globExcludePatterns, threshold, concurrentFiles, tmpDirPath) {
  return checkFlowStatus(flowCommandPath, projectDir, tmpDirPath).then(function (flowStatus) {
    var drainQueue = function () {
      var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3() {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                if (process.env.VERBOSE) {
                  console.log('Wait for ' + waitForCollectedDataFromFiles.length + ' queued files.');
                }
                // Wait the queued files.
                _context3.next = 3;
                return Promise.all(waitForCollectedDataFromFiles);

              case 3:
                // Empty the collected Data From files queue.
                waitForCollectedDataFromFiles = [];

              case 4:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function drainQueue() {
        return _ref4.apply(this, arguments);
      };
    }();

    var now = new Date();
    var coverageGeneratedAt = now.toDateString() + ' ' + now.toTimeString();

    var annotationSummary = {
      passed: false,
      flowFiles: 0,
      flowWeakFiles: 0,
      noFlowFiles: 0,
      totalFiles: 0
    };

    var coverageSummaryData = {
      threshold: threshold,
      covered_count: 0, uncovered_count: 0, // eslint-disable-line camelcase
      percent: 0,
      generatedAt: coverageGeneratedAt,
      flowStatus: flowStatus,
      flowAnnotations: annotationSummary,
      files: {},
      globIncludePatterns: globIncludePatterns,
      globExcludePatterns: globExcludePatterns,
      concurrentFiles: concurrentFiles
    };

    // Remove the source attribute from all ucovered_locs entry.
    function cleanupUncoveredLoc(loc) {
      delete loc.start.source;
      delete loc.end.source;
      return loc;
    }

    var waitForCollectedDataFromFiles = [];

    function collectCoverageAndGenerateReportForGlob(globIncludePattern) {
      var _this = this;

      return (0, _promisified.glob)(globIncludePattern, { cwd: projectDir, root: projectDir }).then(function () {
        var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(files) {
          var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _loop, _iterator, _step, _ret;

          return _regenerator2.default.wrap(function _callee4$(_context5) {
            while (1) {
              switch (_context5.prev = _context5.next) {
                case 0:
                  _iteratorNormalCompletion = true;
                  _didIteratorError = false;
                  _iteratorError = undefined;
                  _context5.prev = 3;
                  _loop = _regenerator2.default.mark(function _loop() {
                    var filename;
                    return _regenerator2.default.wrap(function _loop$(_context4) {
                      while (1) {
                        switch (_context4.prev = _context4.next) {
                          case 0:
                            filename = _step.value;

                            if (!(globExcludePatterns.find(function (pattern) {
                              return (0, _minimatch2.default)(filename, pattern);
                            }) !== undefined)) {
                              _context4.next = 4;
                              break;
                            }

                            if (process.env.VERBOSE) {
                              console.log('Skip ' + filename + ', matched excluded pattern.');
                            }
                            return _context4.abrupt('return', 'continue');

                          case 4:

                            if (process.env.VERBOSE) {
                              console.log('Queue ' + filename + ' flow coverage data collection');
                            }

                            waitForCollectedDataFromFiles.push(collectFlowCoverageForFile(flowCommandPath, flowCommandTimeout, projectDir, filename, tmpDirPath).then(function (data) {
                              /* eslint-disable camelcase */
                              coverageSummaryData.covered_count += data.expressions.covered_count;
                              coverageSummaryData.uncovered_count += data.expressions.uncovered_count;
                              data.percent = getCoveredPercent(data.expressions);

                              if (!data.filename) {
                                throw new Error('Unxepected missing filename from collected coverage data');
                              }

                              coverageSummaryData.files[data.filename] = data;

                              data.expressions.uncovered_locs = data.expressions.uncovered_locs.map(cleanupUncoveredLoc);
                              /* eslint-enable camelcase */
                            }));

                            // If we have collected at least `concurrentFiles` number of files,
                            // wait the queue to be drained.

                            if (!(waitForCollectedDataFromFiles.length >= concurrentFiles)) {
                              _context4.next = 9;
                              break;
                            }

                            _context4.next = 9;
                            return drainQueue();

                          case 9:
                          case 'end':
                            return _context4.stop();
                        }
                      }
                    }, _loop, _this);
                  });
                  _iterator = files[Symbol.iterator]();

                case 6:
                  if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
                    _context5.next = 14;
                    break;
                  }

                  return _context5.delegateYield(_loop(), 't0', 8);

                case 8:
                  _ret = _context5.t0;

                  if (!(_ret === 'continue')) {
                    _context5.next = 11;
                    break;
                  }

                  return _context5.abrupt('continue', 11);

                case 11:
                  _iteratorNormalCompletion = true;
                  _context5.next = 6;
                  break;

                case 14:
                  _context5.next = 20;
                  break;

                case 16:
                  _context5.prev = 16;
                  _context5.t1 = _context5['catch'](3);
                  _didIteratorError = true;
                  _iteratorError = _context5.t1;

                case 20:
                  _context5.prev = 20;
                  _context5.prev = 21;

                  if (!_iteratorNormalCompletion && _iterator.return) {
                    _iterator.return();
                  }

                case 23:
                  _context5.prev = 23;

                  if (!_didIteratorError) {
                    _context5.next = 26;
                    break;
                  }

                  throw _iteratorError;

                case 26:
                  return _context5.finish(23);

                case 27:
                  return _context5.finish(20);

                case 28:
                  if (!(waitForCollectedDataFromFiles.length > 0)) {
                    _context5.next = 31;
                    break;
                  }

                  _context5.next = 31;
                  return drainQueue();

                case 31:
                  return _context5.abrupt('return', files);

                case 32:
                case 'end':
                  return _context5.stop();
              }
            }
          }, _callee4, _this, [[3, 16, 20, 28], [21,, 23, 27]]);
        }));

        return function (_x9) {
          return _ref5.apply(this, arguments);
        };
      }());
    }

    return Promise.all(globIncludePatterns.map(collectCoverageAndGenerateReportForGlob)).then(function () {
      coverageSummaryData.percent = getCoveredPercent(coverageSummaryData);
      coverageSummaryData.flowAnnotations = summarizeAnnotations(coverageSummaryData);

      return coverageSummaryData;
    });
  });
}
